hopsworks-api/src/main/java/io/hops/hopsworks/api/provenance/ProvenanceCleaner.java

```diff
@@ -0,0 +1,143 @@
+import org.javatuples.Pair;
+
+import javax.ejb.EJB;
+import javax.ejb.Schedule;
+import javax.ejb.Singleton;
+import javax.ejb.Timer;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Singleton
+@TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
+public class ProvenanceCleaner {
+  private final static Logger LOGGER = Logger.getLogger(ProvenanceCleaner.class.getName());
+  
+  @EJB
+  private Settings settings;
+  @EJB
+  private ProvElasticController client;
+  @EJB
+  private ProjectFacade projectFacade;
+  @EJB
+  private InodeFacade inodeFacade;
+  
+  private String lastIndexChecked = "";
+  
+  // Run once per hour
+  @Schedule(persistent = false, hour = "*", minute = "*")
+  public void execute(Timer timer) {
+    int cleanupSize = settings.getProvCleanupSize();
+    int archiveSize = settings.getProvArchiveSize();
+    if(archiveSize == 0) {
+      return;
+    }
+    try {
+      Pair<Integer, String> round = archiveRound(lastIndexChecked, cleanupSize);
+      LOGGER.log(Level.INFO, "cleanup round - idx cleaned:{0} from:{1} to:{2}",
+        new Object[]{round.getValue0(), lastIndexChecked, round.getValue1()});
+      lastIndexChecked = round.getValue1();
+    } catch (ProvenanceException | ElasticException e) {
+      LOGGER.log(Level.INFO, "cleanup round was not successful - error", e);
+    }
+  }
+  
+  private Pair<Integer, String> archiveRound(String nextToCheck, Integer limitIdx)
+    throws ProvenanceException, ElasticException {
+    String[] indices = getAllIndices();
+    
+    int cleaned = 0;
+    String nextToCheckAux = "";
+    for(String indexName : indices) {
+      if(cleaned > limitIdx) {
+        nextToCheckAux = indexName;
+        break;
+      }
+      if(indexName.compareTo(nextToCheck) < 0) {
+        continue;
+      }
+      Project project = getProject(indexName);
+      if(project == null) {
+        LOGGER.log(Level.INFO, "deleting prov index:{0} with no corresponding project", indexName);
+        deleteProvIndex(indexName);
+        cleaned++;
+        continue;
+      }
+    }
+    return Pair.with(cleaned, nextToCheckAux);
+  }
+  
+  private Project getProject(String indexName) throws ProvenanceException {
+    int endIndex = indexName.indexOf(Settings.PROV_FILE_INDEX_SUFFIX);
+    String sInodeId = indexName.substring(0, endIndex);
+    Long inodeId;
+    try {
+      inodeId = Long.parseLong(sInodeId);
+    }catch(NumberFormatException e) {
+      throw new ProvenanceException(RESTCodes.ProvenanceErrorCode.INTERNAL_ERROR, Level.WARNING,
+        "error extracting project from prov index name - format error", e.getMessage(), e);
+    }
+    Inode inode = inodeFacade.findById(inodeId);
+    if(inode == null) {
+      return null;
+    }
+    Project project = projectFacade.findByInodeId(inode.getInodePK().getParentId(), inode.getInodePK().getName());
```

Comment:
What's the point of making a query to the database when you only care if the result is null?

Reply:
notice the ProvenaceCleaner name. This one cleans elastic indices for projects that have been deleted.